docs/testing.md

```diff
@@ -0,0 +1,130 @@
+
+### Common Rust code between `app` and `coordinator`
+
+There will be a number of crates that will be used as dependencies in both `app` and `coordinator`.
+
+### No initial effort to keep different versions of `app` and `coordinator` compatible
+
+The `coordinator` will not be backwards-compatible with older versions of the `app` until later in the development of 10101.
+This means that we will not devote effort to test backwards compatibility between these two systems.
+
+### The `coordinator` is treated as a single, multi-purpose system
+
+In the past, we have modularised `coordinator`-like systems into several components for architectural and strategic reasons.
+That may very well still be the case for 10101, but at this stage we lack the insight to split this system.
+Once that design has been completed we can revise this document accordingly.
+
+### Common `10101-lightning-node` dependency between `app` and `coordinator`
+
+This `10101-lightning-node` is one of the aforementioned Rust crates shared between these two systems.
+It represents the custom Lightning Node built using (a fork of) `rust-lightning` and `rust-dlc`.
+
+### The `app`'s state will be controlled by its Rust backend
+
+Thanks to this we should be able to write system tests against the `app`'s API without involving the frontend.
+
+### The architecture will be similar to this component diagram
+
+This diagram does include some of the assumptions listed above:
+
+![Assumed component diagram](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.githubusercontent.com/get10101/10101/d7619613324df3ce5c0f0a35d97bf24323f9515b/docs/diagrams/testing-component-diagram.puml)
+
+## Strategy
+
+In this section we describe the strategy that we recommend to follow when writing tests for this project.
+Each subsection corresponds to a section of the codebase or a combination thereof.
+
+### Common crates, `app` backend and `coordinator`
+
+All this code is written in Rust.
+
+#### Unit tests
+
+Good unit testing patterns apply.
+In particular, we strive for these tests to be small, structured[^1] and fast.
+Unit tests will ideally only be written for pure functions.
+
+They can be handy as API snapshot tests. For example, to ensure that a `serde` implementation doesn't change.
+
+#### Integration tests
+
+These are tests that depend on multiple components (perhaps across module boundaries).
+Additionally, this kind of tests usually involve mocked state.
+
+We are aiming to not spend too much time maintaining Rust integration tests, in favour of more comprehensive tests.
+
+### The `app` frontend
+
+This includes Flutter code (written in Dart) and the generated `flutter_rust_bridge`.
+In practice, there shouldn't be a strong need to write tests for the `flutter_rust_bridge` code we write (Rust) since it wil be a thin layer.
+
+#### Unit tests
+
+We can apply similar principles as the ones mentioned for Rust code.
+
+#### Widget tests
```

Comment:
In my point of view widget tests are also unit tests.

Reply:
Fair enough. I think we were going off Flutter's own documentation, but it's true that more generally they belong to another category.

Reply:
It depends on your definition of a unit test - if we talk about one distinct atomic unit the widget test is more on an integration test level because it might mock a whole lifecycle of  the widget and it's behavior throughout. I guess it depends how to write that test.

Here's the docs we read when writing this document: https://docs.flutter.dev/testing#widget-tests